src/gurt/dyn_hash.c

```diff
@@ -0,0 +1,1283 @@
+/*
+ * (C) Copyright 2016-2020 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. 8F-30005.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * This file is part of cart, it implements the dynamic hash table functions.
+ */
+#define D_LOGFAC	DD_FAC(mem)
+#include <gurt/common.h>
+#include <gurt/list.h>
+#include <gurt/hash.h>
+#include <gurt/dyn_hash.h>
+
+#define DYNHASH_SIPBITS 6
+#define DYNHASH_BUCKET	(1 << DYNHASH_SIPBITS)
+#define DYNHASH_VECTOR	64
+#define DYNHASH_MAGIC   0xab013245
+
+#define _le64toh(x) (x)
+
+typedef struct dh_field {
+	uint64_t	siphash;
+	void		*record;
+} dh_field_t;
+
+typedef struct dh_bucket {
+	unsigned char	counter;
+	pthread_mutex_t	mtx;
+	dh_field_t	field[DYNHASH_BUCKET];
+} dh_bucket_t;
+
+
+#define ROTATE(x, b) (uint64_t)(((x) << (b)) | ((x) >> (64 - (b))))
+
+#define HALF_ROUND(a, b, c, d, s, t) \
+	a += b; c += d;                  \
+	b = ROTATE(b, s) ^ a;            \
+	d = ROTATE(d, t) ^ c;            \
+	a = ROTATE(a, 32)
+
+#define DOUBLE_ROUND(v0, v1, v2, v3)         \
```

Comment:
(style)  Macros with multiple statements should be enclosed in a do - while loop